keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1806 @@
+resource usage of the pod is usually lower than the sum of the individual peak
+usages. By establishing pod-level limits based on this aggregate demand,
+organizations can enhance resource efficiency and avoid unnecessary
+over-allocation.
+
+For instance, in an **AI/ML Batch Processing Workload** involving tightly coupled
+containers that execute tasks on large datasets, resource demands can vary
+significantly based on the size of the data and the complexity of the operations.
+Different stages of batch processing might experience peaks at different times.
+For example, data cleaning may require more memory when handling complex
+datasets, whereas data transformation could be less resource-intensive. If
+resources are allocated at the container level based on peak estimates, this can
+lead to scenarios where one container consumes its allocated resources while
+others remain underutilized, resulting in inefficiencies.
+
+Moreover, manually tuning requests and limits for each container can be
+time-consuming and complex. By defining pod-level limits, all containers can
+dynamically share the designated resources. For example, when one container is
+performing a resource-intensive task, it can leverage the unused resources of
+other containers within the pod. This approach not only improves overall resource
+utilization but can also lead to cost savings in cloud environments where
+resource allocation impacts billing.
+
+### Notes/Constraints/Caveats
+
+1. cgroupv1 has been moved to maintenance mode since Kubernetes version 1.31.
+   Hence this feature will be supported only on cgroupv2. Kubelet will fail to
+   admit pods with pod-level resources on nodes with cgroupv1.
+2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
+effect on existing deployments. Only deployments that explicitly require this
+functionality should turn it on by setting the relevant resource section at
+pod-level in the Pod specification.
+
+### Risks and Mitigations
+
+1. Since Pod Level Resource Specifications is an **opt-in** feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+
+2. While this feature doesn't alter the existing cgroups structure, it does
+   change how pod-level cgroup values are determined. Currently, Kubernetes
+   derived these values from the container-level cgroup settings. However, with
+   Pod Level Resource Specifications enabled, pod-level cgroup settings will be
+   directly set based on the values specified in the pod's resource spec stanza,
+   if set. This change in behavior could potentially affect:
+
+    * Workloads or tools that rely on reading cgroup values: This means that any
```

Comment:
We could look into projecting pod level values as well into the container either as a read only fs or through downward API.

Reply:
For the runtimes relying on container-level values, do they rely only on limits? If yes, the KEP talks about setting the pod-level limits in cgroups. Will that suffice?

Reply:
Yes, that should work :+1: